경북대 BE_탁정민 2주차 과제 (2단계 + 3단계)

xkrwjdals opened the pull request:
# 이전 피드백 반영에 대해
우선 JWT라는 것이나 토큰을 사용해 인증을 구현하는 것에 대해 찾아보고 배우고 있는 중이라 천천히 1단계 피드백을 반영하도록 하겠습니다!

우선 DTO model에 대해서는 `record`를 사용하는 것으로 바꾸었습니다! 그리고 `ProductOperation`도 `ProductRepository`로 리팩토링 했습니다.

# 2단계 과제 질문
## `POST` request 으로 로그인
2단계 과제 예시를 보면 로그인을 할 때 email(혹은 id)와 password 입력값과 함께 `POST` 요청을 보내면 이에 대한 응답으로 서버에서 Token(JWT)를 반환하는 것을 알 수 있습니다. Token을 생성하는 작업이 이루어지기 때문에 `POST` 요청을 사용하는 것인지 혹은 이런 것에 대한 암묵적인 약속이 있는 것인지 궁금합니다. 물론 `POST` 말고 다른 요청을 사용하여도 동작에 문제는 없겠지만, 무언가를 생성하는 행위이기 때문에 `POST` 를 사용하는 것인가요?

## Exception
`AuthController`에서 로그인을 처리하는 부분에서 try catch문을 사용하여 정상적으로 동작하지 않았을 때 예외가 발생될 수 있도록 설정했습니다. 혹시 1주차 과제 피드백에서 말씀하셨던 `GlobalExceptionHandler`에서 다른 예외를 처리하는 방법이 `AdminController` 혹은 `ProductController`에서 `try catch`문을 사용하여 예외를 발생시킨 뒤, `GlobalExceptionHandler`에서 `@ExceptionHandler` 어노테이션을 사용하여 처리할 수 있도록 구현하면 될까요?
그리고 해당 방법이 정석적인 방법일까요?

# 3단계 과제
## 토큰은 어디에?
`POST /api/register`로 email과 password를 담은 `Member`객체가 DB에 저장 -> `POST /api/login`으로 email과 password가 DB에 있는 지 확인 후 존재한다면, JWT를 리턴 -> 그 다음 `GET /wishlist` 요청을 보내면 헤더의 Authorization으로 JWT가 포함되어 서버에 전송되는 것이 맞나요? 

저는 Postman으로 로그인 요청을 보내서 Response로 토큰을 반환받아 이를 헤더 값에 추가하고, 데이터 요청 + JWT가 이루어질 수 있도록했는데, Postman으로 요청을 보내는 것이 아닌 프론트엔드 측에서는 JWT 토큰을 어떻게 활용하나요?

## Repository와 Service
Repository는 DB와의 연결만 구현하고, Service에서는 CRUD의 구현을 담당한다고 알고 있습니다. Repository에서 JdbcTemplate를 사용하여 DB와 데이터 전달을 담당하고 Service에서는 Repository를 생성자 주입으로 의존성을 주입해주고, CRUD를 구현하는 메서드에 Repository의 메서드를 호출하는 방식으로 구현하는게 맞나요? 그렇다면 1주차 과제에서 만든 `MemberRepository`와 `ServiceRepository`는 2개를 지금 한꺼번에 처리하고 있지 않은지..?

## Repository와 Service의 분리
위에서 나온 질문처럼 Repository와 Service를 구분해보자고 생각해서 아래와 같이 코드를 작성하였습니다. 그러자 DB와 정상적인 연결이 되지 않는 문제가 발생했습니다. 그 이유가 궁금합니다. 
### `WishListRepository.java`
```java
package gift.service;

import gift.model.WishList;
import gift.model.WishListDTO;
import java.util.List;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.jdbc.core.BeanPropertyRowMapper;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Repository;

@Repository
public class WishListRepository {

    private final JdbcTemplate jdbcTemplate;
    private static final Logger logger = LoggerFactory.getLogger(WishListService.class);


    public WishListRepository(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }

    public void createWishList(WishListDTO wishListDTO) {
        String sql = "INSERT INTO wishlist (email, product_name, quantity) VALUES (?, ?, ?)";
        jdbcTemplate.update(sql, wishListDTO.email(), wishListDTO.product_name(), wishListDTO.quantity());
    }

    public List<WishList> getAllWishList() {
        String sql = "SELECT * FROM wishlist";
        return jdbcTemplate.query(sql, new BeanPropertyRowMapper<>(WishList.class));
    }

    public List<WishList> getWishListByEmail(String email) {
        String sql = "SELECT * FROM wishlist WHERE email = ?";
        logger.info("Querying database for wish list with email: {}", email);
        return jdbcTemplate.query(sql, new BeanPropertyRowMapper<>(WishList.class), email);
    }

    public boolean deleteWishList(String email, String product_name) {
        String sql = "DELETE FROM wishlist WHERE email = ? AND product_name = ?";
        return jdbcTemplate.update(sql, email, product_name) > 0;
    }
}
``` 
### `WishListService.java`
```java
package gift.service;

import gift.model.WishList;
import gift.model.WishListDTO;
import java.util.List;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

@Service
public class WishListService {
    private final WishListRepository wishListRepository;
    private static final Logger logger = LoggerFactory.getLogger(WishListService.class);

    public WishListService(WishListRepository wishListRepository) {
        this.wishListRepository = wishListRepository;
    }

    public void createWishList(String email, WishListDTO wishListDTO) {
        wishListRepository.createWishList(wishListDTO);
    }

    public List<WishList> getAllWishList() {
        return wishListRepository.getAllWishList();
    }

    public List<WishList> getWishListByEmail(String email) {
        logger.info("Fetching wish list for email: {}", email);
        return wishListRepository.getWishListByEmail(email);
    }

    public boolean deleteWishList(String email, String product_name) {
        boolean deleted = wishListRepository.deleteWishList(email, product_name);
        if (deleted) {
            logger.info("Successfully Deleted From DB");
            return true;
        }
        logger.info("Cannot Deleted From DB");
        return false;
    }
}
``` 

### `WishListController.java`
```java
package gift.controller;

import gift.model.WishList;
import gift.model.WishListDTO;
import gift.security.JwtUtil;
import gift.service.WishListService;
import jakarta.servlet.http.HttpServletRequest;
import java.util.List;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpRequest;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/wishlist")
public class WishListController {
    private final WishListService wishListService;
    private final JwtUtil jwtUtil;
    private static final Logger logger = LoggerFactory.getLogger(WishListController.class);

    public WishListController(WishListService wishListService, JwtUtil jwtUtil) {
        this.wishListService = wishListService;
        this.jwtUtil = jwtUtil;
    }

    @GetMapping
    public ResponseEntity<List<WishList>> getAllWishList() {
        List<WishList> wishlists = wishListService.getAllWishList();
        return ResponseEntity.ok(wishlists);
    }
    @GetMapping("/{email}")
    public ResponseEntity<List<WishList>> getWishList(@PathVariable String email, HttpServletRequest request) {
        String authHeader = request.getHeader("Authorization");

        if (authHeader == null || !authHeader.startsWith("Bearer ")) {
            logger.error("Authorization header is missing or invalid");
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build(); // Unauthorized
        }

        String token = authHeader.substring(7);
        String tokenEmail;

        try {
            tokenEmail = jwtUtil.extractEmail(token);
        } catch (Exception e) {
            logger.error("Error extracting email from token", e);
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build(); // Unauthorized
        }

        logger.info("URL Email: {}, Token Email: {}", email, tokenEmail);

        if (!email.equals(tokenEmail)) {
            logger.warn("Emails do not match: URL Email: {}, Token Email: {}", email, tokenEmail);
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build(); // Forbidden
        }
        List<WishList> wishLists = wishListService.getWishListByEmail(email);
        return ResponseEntity.ok(wishLists);
    }

    @PostMapping("/{email}")
    public ResponseEntity<?> createWishList(@PathVariable String email, HttpServletRequest request, @RequestBody WishListDTO wishListDTO) {
        String authHeader = request.getHeader("Authorization");

        if (authHeader == null || !authHeader.startsWith("Bearer ")) {
            logger.error("Authorization header is missing or invalid");
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build(); // Unauthorized
        }

        String token = authHeader.substring(7);
        String tokenEmail;

        try {
            tokenEmail = jwtUtil.extractEmail(token);
        } catch (Exception e) {
            logger.error("Error extracting email from token", e);
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build(); // Unauthorized
        }

        logger.info("URL Email: {}, Token Email: {}", email, tokenEmail);

        if (!email.equals(tokenEmail)) {
            logger.warn("Emails do not match: URL Email: {}, Token Email: {}", email, tokenEmail);
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build(); // Forbidden
        }
        return ResponseEntity.status(HttpStatus.CREATED).body(wishListDTO);
    }

    @DeleteMapping("/{email}/{product_name}")
    public ResponseEntity<?> deleteWishList(@PathVariable String email, @PathVariable String product_name, HttpServletRequest request) {
        String authHeader = request.getHeader("Authorization");

        if (authHeader == null || !authHeader.startsWith("Bearer ")) {
            logger.error("Authorization header is missing or invalid");
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build(); // Unauthorized
        }

        String token = authHeader.substring(7);
        String tokenEmail;

        try {
            tokenEmail = jwtUtil.extractEmail(token);
        } catch (Exception e) {
            logger.error("Error extracting email from token", e);
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build(); // Unauthorized
        }

        logger.info("URL Email: {}, Token Email: {}", email, tokenEmail);

        if (!email.equals(tokenEmail)) {
            logger.warn("Emails do not match: URL Email: {}, Token Email: {}", email, tokenEmail);
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build(); // Forbidden
        }

        boolean deleted = wishListService.deleteWishList(email, product_name);
        if (deleted) {
            logger.info("Service Layer work correctly");
            return ResponseEntity.noContent().build();
        }
        logger.info("Service Layer Error");
        return ResponseEntity.notFound().build();
    }

}
``` 

## DB 생성?
`Member` 객체가 members DB에 저장될 때, (Member id, 상품의 이름, 갯수, 가격)을 저장하는 wishlist 라는 DB를 생성하는 것을 생각했으나, ChatGPT가 권장하지 않는다는 답변을 받았습니다. 그래서 wishlist DB에
id(Member 객체의 id가 아닌 wishlist의 id), email, product_name, quantity
를 저장하게끔 해서 email을 통해 특정 사용자의 위시 리스트들을 불러올 수 있게 했습니다. 이것이 맞는 방법인지 궁금합니다. 동적으로 DB를 생성하는 것이 왜 권장되지 않고 할 수 있는 방법은 없는지도 궁금합니다.

xkrwjdals committed:
feat: Apply Bean Validation in Spring

- Add 'spring-boot-starter-validation' dependency
- Add @NotEmpty and @Size for validation
- Add @Valid annotation in Service(ProductOperation.java) and Controller(ProductController.java)

xkrwjdals committed:
fix: Correct validation annotations for Controller and Service

xkrwjdals committed:
fix: Correct validation annotations

- String and Collection type can use @NotEmpty only
- integer and long type can use @NotNull only

xkrwjdals committed:
feat: Add regular expression for special symbols validation

xkrwjdals committed:
feat: Add custom constraint with custom annotation and validator

- Check word "카카오" in name field value

xkrwjdals committed:
feat: Add GlobalExceptionHandler.java and error.html

- Handle Bean validation exceptions by GlobalExceptionHandler
- Provide invalid values and reasons to client in error.html

xkrwjdals committed:
docs: Update README.md

xkrwjdals committed:
feat: Add error page thymeleaf template by product-edit.html

- Show error messages from GlobalExceptionHandler
- Add HiddenHttpMethodFilter Bean for PUT request in Application.java

xkrwjdals committed:
fix: Delete HiddenHttpMethodFilter Bean

- Activate hidden method filter in application.properties

xkrwjdals committed:
fix: Apply @ResponseStatus HttpStatus.NOTFOUND in GlobalExceptionHandler

xkrwjdals committed:
fix: Apply @ResponseStatus HttpStatus.BAD_REQUEST in GlobalExceptionHandler

xkrwjdals committed:
docs: Update README.md

xkrwjdals committed:
feat: Add Member and MemberDTO class

- Implement MemberDTO class within password and email

xkrwjdals committed:
feat: Implement MemberRepository

- Connect with DB by JdbcTemplate

xkrwjdals committed:
feat: Implement token authorization by JWT

- Add implementation jjwt-api in build.gradle
- Implement generateToken and method for Controller

xkrwjdals committed:
feat: Implement authorization controller

- Implement POST request by createAuthenticationToken
- Add schema.sql with members TABLE

xkrwjdals committed:
feat: Implement login in AuthController

- Return JWT token when login success
- Fix Member class and MemberRepository correctly

xkrwjdals committed:
docs: Update README.md

xkrwjdals committed:
fix: Apply record to DTO models(ProductDTO, MemberDTO)

- Change name of Service from ProductOperation.java to ProductRepository.java
- Apply optimal import to all files

xkrwjdals committed:
fix: Change method name in AuthController.java

xkrwjdals committed:
fix: Resolve merge conflicts

xkrwjdals committed:
docs: Update README.md

xkrwjdals committed:
feat(Model): Add WishList domain model and WishListDTO

xkrwjdals committed:
feat(Repository): Implement WishListRepository with getWishListByEmail

- Replace HashMap to product name and quantity in WishList model and DTO model
- Implement getWishListByEmail method
- Add wishlist TABLE in schema.sql

xkrwjdals committed:
feat: Implement create wishlist

- Implement @PostMapping in Controller
- Implement Create wishlist in Service and Repository

xkrwjdals committed:
feat: Implement delete wishlist

- Implement @DeleteMapping in Controller
- Implement Delete wishlist in Service and Repository

xkrwjdals committed:
fix: Delete WishListService.java

- Combine to WishListRepository
- Modify createWishList method to connect DB correctly

xkrwjdals committed:
feat: Implement change product quantity of wishlist

- Add @PutMapping in WishListController
- Add updateWishList method in WishListRepository

jadie0319 commented on src/main/java/gift/controller/AdminController.java:
```diff
@@ -23,14 +22,14 @@
-    public AdminController(ProductOperation productOperation) {
-        this.productOperation = productOperation;
+    private final ProductRepository productRepository;
+    public AdminController(ProductRepository productRepository) {
```
나중을 위해서 Controller 에서 Service 를 의존하게 만들어주는게 좋아요.
Controller 에서 Repository 의 반환값을 직접 다루게 되면 Controller 가 많은 책임을 맡게 됩니다. 지금 deleteProduct의 if 문과 같은 로직이 더 생기게 될지도 모릅니다.

Controller 는 사용자가 원하는 결과값을 전달해주는 역할만 할 수 있게
Repository 로 부터 오는 값은 Servicve 영역에서 다루게 하면 어떨까요?

xkrwjdals replied:
상세한 피드백 감사합니다! `ProductService.java` 를 새로 만들어 아래 내용을 따르도록 코드를 수정했습니다!

Controller -> Service를 의존, 사용자의 URI 요청을 받는다. 요청에 대한 응답을 View에 반환
Service -> Repository를 의존, 비즈니스 로직을 처리
Repository -> DB와의 연결
제가 이해한 것이 맞을까요?

jadie0319 replied:
넵 맞습니다!

jadie0319 commented on src/main/java/gift/controller/AdminController.java:
```diff
@@ -45,26 +44,26 @@ public ModelAndView showAddPage() {
     @DeleteMapping("/delete/{id}")
     public ResponseEntity<String> deleteProduct(@PathVariable Long id) {
-        boolean check = productOperation.deleteProduct(id);
+        boolean check = productRepository.deleteProduct(id);
```
service 를 통해 deleteProduct 로직을 구현하게 되면 아래 if 로직을 없앨 수 있을 것 같아요 :)
ExceptionHandler 사용을 고려해보시는 것도 좋을 것 같아요

jadie0319 commented on src/main/java/gift/controller/AuthController.java:
```diff
@@ -0,0 +1,41 @@
+        this.jwtUtil = jwtUtil;
+    }
+
+    @PostMapping("/register")
```
api url 이 명확하지 않은 것 같아요. 무엇을 register 하는 api 일까요?

xkrwjdals replied:
API URL을 잘못 지은것 같아, `/api/members`로 `@RequestMapping` 한 뒤 `@RequestBody`에 `MemberDTO` 객체의 필드 변수 값을 JSON type으로 `POST` 요청을 보내는 경우, members DB에 저장될 수 있도록 수정했습니다!

그리고 `/api/members/login` 으로 `POST` 요청을 보내는 경우에는 token을 응답으로 줄 수 있도록 했는데, 'login' 이라는 단어도 행위를 나타내는 단어라 부적절하다고 생각합니다. 어떻게 수정하면 좋을까요?

jadie0319 replied:
회원의 동록, 삭제, 조회, 수정 과 같은 특정 자원의 REST API 설계는 '/' 를 depth 로 하여 표현하곤 해요.
그런데 login, signup, logout 과 같은 특정 자원은 아니고 행위를 표현할 때는 그냥 사용하기도 합니다.
엄격하게 구현하려면 token 이라는 자원으로 표현할 수도 있겠는데 저도 와닿지가 않네요...
API 는 Application Programming Interface 의 약자예요. 결국 Interface 이고 곧 약속, 규약과 같은 거죠.
어느정도 지켜야 하는 규칙은 있지만 약속과 규약은 이해관계자 합의 하에 만들면 된다고 생각합니다!
정민님이 생각하는 것을 url 로 표현했으면 그걸로 됐다고 생각합니다 :)

register 같은 경우는 없애도 된다 판단해서 말씀 드렸고, login 은 그냥 둬도 될 것 같아서 말씀을 안드렸어요.
한번 고민해보시라고 언급이라도 했으면 좋았겠네요 ㅎㅎ

jadie0319 commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -20,21 +20,21 @@
 public class ProductController {
 
-    private final ProductOperation productOperation;
+    private final ProductRepository productRepository;
```
AdminController 피드백과 동일합니다. Service 로직을 참조하게 해주시는게 좋을 것 같아요.

jadie0319 commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -20,21 +20,21 @@
     public ResponseEntity<Product> getProductById(@PathVariable Long id) {
-        Product product = productOperation.getProductById(id);
+        Product product = productRepository.getProductById(id);
         if (product != null) {
             return ResponseEntity.ok(product);
         }
```
sproductService 에서 처리하여 if 문을 없애보면 어떨까요?

jadie0319 commented on src/main/java/gift/controller/AuthController.java:
```diff
@@ -0,0 +1,41 @@
+public class AuthController {
+    private final MemberRepository memberRepository;
+    private final JwtUtil jwtUtil;
+    public AuthController(MemberRepository memberRepository, JwtUtil jwtUtil) {
```
service 를 만들게 되면 jwtUtil 도 Controller 에서 사용하지 않아도 될 것 같아요!

jadie0319 commented on src/main/java/gift/controller/WishListController.java:
```diff
@@ -0,0 +1,167 @@
+    @GetMapping("/{email}")
+    public ResponseEntity<List<WishList>> getWishList(@PathVariable String email,
+        HttpServletRequest request) {
+        String authHeader = request.getHeader("Authorization");
+
+        if (authHeader == null || !authHeader.startsWith("Bearer ")) {
+            logger.error("Authorization header is missing or invalid");
+            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build(); // Unauthorized
+        }
+
+        String token = authHeader.substring(7);
+        String tokenEmail;
+
+        try {
+            tokenEmail = jwtUtil.extractEmail(token);
+        } catch (Exception e) {
+            logger.error("Error extracting email from token", e);
+            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build(); // Unauthorized
+        }
+
+        logger.info("URL Email: {}, Token Email: {}", email, tokenEmail);
+
+        if (!email.equals(tokenEmail)) {
+            logger.warn("Emails do not match: URL Email: {}, Token Email: {}", email, tokenEmail);
+            return ResponseEntity.status(HttpStatus.FORBIDDEN).build(); // Forbidden
+        }
```
해당 코드는 중복으로 사용되는 코드네요. 코드 중복을 제거해볼수 있을 것 같아요.
메서드 분리를 하셔도 되는데 수업시간에 다룬 HandlerMethodArgumentResolver 를 사용하면
더 명확하게 기능이 분리될 것 같아요!

xkrwjdals replied:
`AuthArgumentResolver` 라는 이름으로 custom resolver를 만들어 구현해보았습니다!

또한 Service와 Repository를 구분하여 위의 3개의 피드백도 반영했습니다! 확인 부탁드립니다!

jadie0319 commented on src/main/java/gift/security/JwtUtil.java:
```diff
@@ -0,0 +1,41 @@
+
+import java.util.Date;
+
+@Component
```
JwtUtil 클래스를 컴포넌트로 구현하셨네요! 혹시 이유가 있을까요?

관련해서 좋은글이 있네요. 참고 하시면 좋을 것 같아요.
https://velog.io/@idonymyeon/Util-클래스를-써야하나-Component를-써야하나

jadie0319 commented on src/main/java/gift/service/MemberRepository.java:
```diff
@@ -0,0 +1,33 @@
+    public Member createMember(MemberDTO memberDTO) {
+        String sql = "INSERT INTO members (password, email) VALUES (?, ?)";
+        jdbcTemplate.update(sql, memberDTO.password(),memberDTO.email());
+        return getByEmail(memberDTO.email());
```
update 후 다시 조회하는 이유가 있을까요?

jadie0319 commented on src/main/java/gift/service/WishListRepository.java:
```diff
@@ -0,0 +1,53 @@
+        String sql = "INSERT INTO wishlist (email, product_name, quantity) VALUES (?, ?, ?)";
+        jdbcTemplate.update(sql, wishListDTO.email(), wishListDTO.product_name(),
+            wishListDTO.quantity());
+        return getWishListByProductName(wishListDTO.product_name());
```
코멘트에도 달아 드렸는데 다시 조회를 안해도 될것 같은데 어떻게 생각하시나요?

xkrwjdals replied:
맞다고 생각합니다! 1주차에 크게 생각하지 않고 무작정 return 값을 주게끔 `ProductRepository` 코드를 작성했는데, 이번 2주차 과제에서도 `ProductRepository`, `ProductController` 를 참고하여 `WishListRepository`나 `WishListController` 클래스를 만들었다보니 발생한 문제인 것 같습니다!

jadie0319 commented on src/main/java/gift/service/WishListRepository.java:
```diff
@@ -0,0 +1,53 @@
+
+    public WishList getWishListByProductName(String product_name) {
+        String sql = "SELECT * FROM wishlist WHERE product_name = ?";
+        return jdbcTemplate.queryForObject(sql, new BeanPropertyRowMapper<>(WishList.class),
```
queryForObject 메서드 관련 코멘트에 써놨습니다. 참고 부탁드려요!

xkrwjdals replied:
확인했습니다 감사합니다!

jadie0319 commented on src/main/java/gift/model/WishListDTO.java:
```diff
@@ -0,0 +1,5 @@
+package gift.model;
+
+public record WishListDTO(String email, String product_name, long quantity) {
```
변수명엔 카멜케이스를 많이 사용합니다. productName 이 더 좋을 것 같아요 !

jadie0319 reviewed (changes requested):
2주차 2,3단계 미션하시느라 고생하셨습니다!
피드백 반영 부탁드립니다~
질문주신것은 코멘트에 정리해놓겠습니다!

jadie0319 commented:
안녕하세요! 정민님. 
2,3단계 미션 하시느라 고생 하셨습니다! 피드백 참고해주시고 반영해주세요~
질문 드린것 답변 드리겠습니다.

# 2단계 

질문1 : **`POST` request 으로 로그인**

우선 GET 메서드를 사용하면 아이디와 비밀번호를 쿼리스트링으로 전달해줘야 하기 때문에 보안 이슈가 있어요. 그래서 외부에 노출되지 않게 BODY 에 넣어 POST 로 전달합니다.

  그리고 GET 은 서버의 어떤 값을 변경, 생성하면 안됩니다. 하지만 로그인은  토큰이 생성되죠. 그래서 POST 를 사용합니다.

 그렇다고 토큰을 생성하기 때문에 POST 를 사용한다고는 생각 안하셔도 될 것 같아요. 토큰을 생성하기 위해 POST 를 썼다면 /api/token 형태의 url 을 호출할 것 같아요. 

이 질문은 답변이 참 어렵네요 ㅎㅎ;;  login 의 주목적은 입력받은 유저를 검증하는 것이 목적이고 토큰이 부가적으로 생성되기 때문에 POST 를 사용한다. 그리고 보안 목적으로 POST 를 써야 한다 정도로만 알고 계시면 될 것 같습니다.

질문2 : Exception

이게 정석! 이라고 딱 잘라 말씀드리긴 어렵네요. 하지만 스프링을 사용한다면 알아야 하는 방법이고 가장 많이 사용되는 보편적인 방법입니다. 

반드시 try,catch 를 구현하실 필요는 없어요. 예를 들면 아래와 같은 형태의 코드가 있을 수 있어요.

```java
public class ProductService {
	public Product findproductById(Long id) {
		// productRepository.findById 반환값이 Product
		Product product = productRepository.findById(id);
		if (product == null) {
			throw new NotFoundProductException("상품이 없습니다");
		}
		또는
		// productRepository.findById 반환값이 Optional<Product>
		Product product = productRepostiroy.findById(id)
			.orElseThorw(() -> new NotFoundProductException("상품이 없습니다"));
	}
}
```

이런식으로 예외를 발생시킨 후 ExceptionHandler 클래스에서 예외를 처리하는 방법이 있지요. 

그리고 이런 로직은 Controller 의 역할이 아니기 때문에 Service 단에 위치시키시면 좋을 것 같아요~


# 3단계

질문1 : **토큰은 어디에?**

`POST /api/register`로 email과 password를 담은 `Member`객체가 DB에 저장 -> `POST /api/login`으로 email과 password가 DB에 있는 지 확인 후 존재한다면, JWT를 리턴 -> 그 다음 `GET /wishlist` 요청을 보내면 헤더의 Authorization으로 JWT가 포함되어 서버에 전송되는 것이 맞나요?

넵 말씀하신 로직이 이번 미션에서 구현해야 하는 로직입니다! :) 프론트 측에서는 쿠키에 저장된 토큰을 읽어서 보내는 경우가 많아요. 아래 링크 읽어보시면 도움이 되실 것 같아요. 그리고 토큰, 세션, 쿠키 와 같은 키워드로 검색해보시면 좋을 것 같아요!

https://time-map-installer.tistory.com/236

질문2 : **Repository와 Service**

말씀하신 방법 맞습니다. Controller → Service → Repository 라고 생각하시면 됩니다. 1주차 과제에서 MemberRepository, ServiceRepository 가 보이지 않습니다ㅠ 어떤게 궁금하신걸까요??

질문3 : **Repository와 Service의 분리**

service 사용하는 코드로 변경해서 실행했는데 잘됩니다

```java
List<WishList> wishListByEmail = service.getWishListByEmail(email);
List<WishList> wishLists = wishListRepository.getWishListByEmail(email);
```

당시에 어떤 에러가 났는지를 말해주시면 더 좋을것 같습니다.

제가 발견한 에러는 service 로직의 문제가 아니라 repository 에 있었습니다. 아래 에러 메시지 입니다.

```java
org.springframework.dao.IncorrectResultSizeDataAccessException: Incorrect result size: expected 1, actual 2
```

위시리스트를 만들 때 `getWishListByProductName` 메서드를 사용합니다. 이 메서드 안에서 `jdbcTemplate.queryForObject` 를 사용할 때 발생하는 에러인데 쿼리를 통해 조회한 위시리스트가 두개 이상이면 발생합니다. queryForObject 는 쿼리의 결과가 한건만 나와야 하는 메서드 입니다. jdbcTemplate 의 query 를 사용하면 발생 안할 것 같네요.

추가로 위시리스트를 create, update 하고 getWishlistByProductName 을 통해  조회하는 로직이 있는데, 없어도 되는 로직이라 생각하는데 어떻게 생각하시나요? 

질문4 : **DB 생성?**

제가 질문을 잘 이해하지 못하겠어요. 그냥 member 와 wishlist 테이블에 대해 설명해볼게요.

H2 DB 테이블은 어플리케이션이 실행될때 만들어질거에요. member db 의 저장과 wishlist 의 저장은 각각 이루어져야 합니다. 그리고 회원의 wishlist 를 조회할 때 member의 id 로 조회할지 email 로 조회할지는 DB 테이블 설계가 어떻게 되느냐에 따라 달라요.

```java
CREATE TABLE IF NOT EXISTS members (
    id BIGINT AUTO_INCREMENT PRIMARY KEY,
    password VARCHAR(100) NOT NULL,
    email VARCHAR(50) NOT NULL
);

CREATE TABLE IF NOT EXISTS wishlist (
    id BIGINT AUTO_INCREMENT PRIMARY KEY,
    email VARCHAR(50),
    product_name VARCHAR(255),
    quantity BIGINT
);
```

현재의 DB 구조일 때 아래와 같은 회원 데이터가 저장되어 있어요.

- id=1, email=”kkk@gmail.com”
- id=2, email=”kkk@gmail.com”

회원의 wishlist 를 조회하려면 무엇으로 조회해야 할까요? email 로 조회하면 id가 1,2인 회원들의 위시리스트를 모두 조회할 것 같지 않나요?

미션의 요구사항을 충족하는 특정 회원의 위시리스트만 조회하려면 회원을 식별할 수 있는 고유한 키값인 회원의 id 로 조회해야 할 것 같아요. 그리고 wishlist 의 테이블도 email 대신에 회원의 id 를 저장하는 식으로 변경되어야 할 것 같아요.

이렇게 된다면 위시리스트 조회 api url 도 /api/wishlist/{memberId} 와 같은 식으로 변경되어야 할 것 같아요~

jadie0319 commented on src/main/java/gift/controller/WishListController.java:
```diff
@@ -0,0 +1,167 @@
+        return ResponseEntity.ok(wishlists);
+    }
+
+    @GetMapping("/{email}")
```
api url 이 나타내는 것이
특정 이메일을 가진 위시리스트를 조회하는 api 라는 의미 일까요?

xkrwjdals replied:
맞습니다!

jadie0319 commented on src/main/java/gift/controller/WishListController.java:
```diff
@@ -0,0 +1,167 @@
+        return ResponseEntity.ok(wishLists);
+    }
+
+    @PostMapping("/{email}")
```
/wishlist/{email}
이라고 하면 조금 어색한것 같아요. rest api 는 / 단위로 특정 리소스를 식별합니다.
wishlist 에 상품을 추가하는 로직이니
/wishlist/product 라고 하면 어색할까요? 위시리스트 상품 등록에 필요한 데이터는 dto 로 받고요!

xkrwjdals replied:
생각을 해보니 `/wishlist/product` 로 수정하는 것이 맞는 거 같습니다!

xkrwjdals committed:
feat(service): Create ProductService and integrate with controller

- create ProductService to handle business logic
- modifiy ProductController to use ProductService
- add ProductRepository to manage DB interactions
= add RepositoryException to handle exception

xkrwjdals committed:
feat(exception): Add UserNotFoundException when login

xkrwjdals committed:
fix(repository): Change return value from Product to void

xkrwjdals committed:
feat(service): create member service and integrate with controller

- created MemberService to handle business logic
- modified MemberController to use MemberService
- added MemberService to manage DB interactions

xkrwjdals committed:
feat: Implement custom resolver to handle JWT authorization

- Add model named by AuthInfo

xkrwjdals committed:
feat(exception): Implement handleUnauthorizedException method in GlobalExceptionHandler with UnauthorizedException

xkrwjdals committed:
feat(service): create wishlist service and integrate with controller

- created WishListService to handle business logic
- modified WishListController to use WishListService
- modified WishListRepository to manage DB interactions
- change product_name to productName

xkrwjdals commented:
PR 질문 답변 감사합니다!
우선 2단계 과제에서 질문드렸던 부분은 모두 궁금증이 해소된 것 같습니다. 감사합니다!
3단계 과제에서 
## 질문 2
제가 묻고 싶었던 것은 제가 이해한 바가 맞는가? 였습니다. 방금 코드 리뷰 피드백 댓글로도 답을 받았습니다. 감사합니다. `ServiceRepository` 경우 제가 오타를 냈습니다....
## 질문 3
어떠한 에러가 특별히 발생하지는 않았고, Postman을 통해 요청을 보냈을 때는 정상적으로 동작하지만, DB에 업데이트가 되지 않는 문제가 있었습니다. 예를 들면 어떠한 `WishListDTO` 를 wishlist 라는 DB에 추가하는 `POST` 요청을 보냈을 때 에러나 예외가 발생하지는 않았지만 정상적으로 DB에 추가가 되지 않았습니다. 하지만 오늘 피드백을 반영하면서 다시 과제를 수행해보니 정상적으로 잘 작동했습니다. 아마 제가 중간에 뭔가 빼먹은게 있지 않을까 싶습니다. 해결 완료했습니다. 답변 감사합니다!
## 질문 4
제가 궁금했던 점은 
'member1 이라는  객체를 members DB에 추가할 때마다, member1_wishlist 라는 DB가 새로 생성되면 안되는가?' 였습니다. 

추가적으로 고유한 키값인 회원 id로 위시 리스트를 조회하는 것이 맞다고 생각합니다. 처음에 회원가입과 로그인을 하는 기능을 구현할 때, 'email과 password를 사용하여 로그인한다.' 라고 생각해서 email을 사용했던 것 같습니다.

xkrwjdals committed:
feat(model): Add memberId value in WishList models

xkrwjdals committed:
fix: Change POST and GET API url from email to memberId

- Fix AuthInfo and AuthArgumentResolver to use memberId

xkrwjdals committed:
fix: Change PUT and DELETE API url from email to memberId

jadie0319 commented:
'member1 이라는 객체를 members DB에 추가할 때마다, member1_wishlist 라는 DB가 새로 생성되면 안되는가?'

member1_wishlist 라는 테이블이 생긴다고 이해하면 될까요?
아니면 wishlist 라는 테이블에 데이터가 저장된다고 이해하면 될까요?

우선 테이블이 생긴다고 생각하면... 회원이 100만명이면 테이블이 100만개가 더 생기게 됩니다. 
그리고 그 테이블 당 회원 데이터는 몇개 없을 것 같네요.

생각을 하다보니 member 가 생겼을 때, 그 회원의 wishlist 를 담을 수 있는 저장공간을 자동으로 생성하게 하는것?? 
이러한 것을 의도하셨던거라는 추측이 드네요.

이것을 의도하셨던 거라면 우선 고민해 봐야 할것은

member 가 생기는 것과 동시에 wishlist 가 꼭 생겨야 하는지를 고려해봐야 하고 (위시리스트에 상품을 담지 않는 회원도 있으니까요 ㅎ)

member 와 wishlist 의 관계도 고려해봐야해요.
wishlist 에 저장되는 데이터에 member_id 를 저장할 수 있다고 상상해보면
member 테이블과 wishlist 의 테이블을 JOIN 하면
member 가 가지고 있는 wishlist 목록을 조회할 수 있어요.
그러면 wishlist  테이블을 동적으로 만들 필요가 없어져요.


음...  도움이 되셨을지 모르겠네요 ㅠ 추가로 설명이 필요하시면 말씀해주세요!

jadie0319 reviewed (approved):
2주차 미션 하시느라 고생 많으셨습니다! 피드백 꼼꼼히 반영해주셔서 감사합니다 :)
피드백 반영해주신건 3주차 미션 때 더 살펴볼게요! 일단 3주차 미션 시작을 위해 머지하겠습니다! 👍 
질문 주셨던 것은 코멘트로 달아놨습니다. 확인해보시고 이해가 잘 안되시거나 추가 설명이 필요하시면 말씀해주세요 :) 3주차도 화이팅 입니다!
